### ISSUE TITLE
Cannot reproduce NN outside of vw

### ISSUE THREAD

**Issue by jp341** [JohnLangford/vowpal_wabbit#1337]
Is there any documentation on the output model parameters for NN?  
As far as I can tell, it seems to be a non-standard, I'm unable to reproduce predicted values for even trivially simple networks.  
Here's an example:
test_data ( i.e. label = input variable)
0.000019 | 0:0.000019 
0.000001 | 0:0.000001 
0.000017 | 0:0.000017 
-0.000011 | 0:-0.000011 
0.000005 | 0:0.000005 
0.000012 | 0:0.000012 
0.000015 | 0:0.000015 
0.000018 | 0:0.000018 

I train a NN with 1 hidden node:
`vw -d test_data --nn 1 --passes 100 -k -c --cache_file test_model.cache -f test_model.mod`

Then invert the hashes & do predictions:
`vw -d test_data -i test_model.mod --invert_hash test_model.read -p test_model.pred`

Here's the model I get:

> :0
> 0:0:-1.06665
> Constant:232120:3.17207e-05
> HiddenBias:29015:-0.3161
> OutputLayer0[1]:156908:0.143475
> OutputLayerConst[1]:156909:1.7179e-05
> OutputWeight[1]:156908:0.143475

Here are the predictions:
0.000018
0.000014
-0.000011
0.000017
-0.000011
0.000004
0.000012
0.000015

OutputLayer0[1] and OutputWeight[1] seem to have the same hash and value, presumably only one of these are used?

Consider the first value.  If I assume "0" is the weight of 0 input in first node, "Constant" is constant input into that node, "HiddenBias" is constant input into final layer of first node (already things are getting non-standard, testing shows that HiddenBias is a vector of length equal to number of hidden nodes), "OutputLayer0[1]" is weight of first node, and OutputLayerConst[1] is a final constant to add to the mix.

Then, I should expect predictions are as follows:

`(tanh(X[0] * "0" + "Constant") + "Hidden Bias") * OutputLayer0[1] + OutputLayerConst[1]`

so, the first row would be:
(tanh(0.000019 * -1.06665 + 3.17207e-5) -0.3161) * 0.143475 + 1.7179e-5 = -0.04533362
And not 0.000018 as the predictions suggested.

I've tried several other arrangements for possible interpretations of these terms but cannot get even close to reproducing the predictions.

Please help!

**Comment by JohnLangford** [comment]
There isn't a specification other than what is in nn.cc which relatively
opaque.
https://github.com/JohnLangford/vowpal_wabbit/blob/master/vowpalwabbit/nn.cc

Why do you need to reproduce?   VW has many io modes so just using it as a
library is pretty viable.

-John

**Comment by arielf** [comment]
Interesting.  After adding `--holdout_off` (missing in the original report, and I believe should be added), I can't get the model to fully converge (to 1e-6 accuracy) on this data-set even after increasing passes to a much higher number (I gave up after 100000 passes)...

Removing `--nn 1` converges perfectly (to 1e-6) with just 2 passes.

When I add `--audit` I see that in some of the examples the sum of gradients of the 0 feature is tiny (XXe-19):
```
-0.000020 0.000018
        :0:1.8e-05:-1.2236@1.60437e-19  Constant:232120:1:2.21964e-06@1.16258e-09
```
I get somewhat better results if I use larger values (e.g. multiply all feature & label values by 1000)

Other things I tried (which didn't improve convergence): `--nn 2`, `--nn 3`, adding a second feature with same value to each example; Avoiding feature `0` (using `1` and `1`,`2`)

**Comment by jp341** [comment]
Thanks both for your responses!  I really appreciate it!

@JohnLangford : I already have my own code for evaluating a neural network, I just want to use VW for calibrating the parameters.  I'm using NN in a real-time setting so want to keep to my own code that I understand.  As you point out, nn.cc is quite opaque.  I spent some hours trying to understand it before giving up.  NNs are simple math just multiply, add and sigmoid, it shouldn't be complicated to reproduce.  I should point out I also wrote an algorithm trying to bruteforce the equation from the output parameters & operators (*,+,sgm) but failed to find anything sensible.  Something very odd is going on indeed.

@arielf : Yes, sorry I forgot the --holdout_off option, I did use it!
It's been noted by others practically that scaling issues can be an obstacle to training NNs, I think that's why layer-wise RBM pre-training got into vogue as apparently that helped scaling.  I read that VW uses a type of SGD that is highly robust to scale so I was excited to try out NN for this reason.

**Comment by JohnLangford** [comment]
It's the optimizations and options that leave the code a bit opaque.

@pmineiro may advise further.

This line: https://github.com/JohnLangford/vowpal_wabbit/blob/master/vowpalwabbit/nn.cc#L164 computes the activations of the hidden layer.

This line: https://github.com/JohnLangford/vowpal_wabbit/blob/master/vowpalwabbit/nn.cc#L271 is computing the output.

To understand what's going on, maybe just adds some printouts to the code?

**Comment by jp341** [comment]
@JohnLangford : thanks for your pointers.

I tried compiling vw on windows (using VS2015 x86 x64 Cross Tools Cmd Prompt) and it failed with 97 errors.  

**Cmd:** msbuild /p:Configuration=Release /p:Platform="x64" vw.sln
**First error:**
```
Project "\vowpal_wabbit-master\vowpalwabbit\vw.sln" (1) is building "\vowpal_wabbit-master\cs\setup
\setup.wixproj" (16) on node 1 (default targets).
vowpal_wabbit-master\cs\setup\setup.wixproj(102,3): error MSB4019: The imported project "C:\Program Files (x86)\MSBuild\Mic
rosoft\WiX\v3.x\Wix.targets" was not found. Confirm that the path in the <Import> declaration is correct, and that the file exists on disk.
Done Building Project "\vowpal_wabbit-master\cs\setup\setup.wixproj" (default targets) -- FAILED.
```

Seems to be failing on c# stuff, shame since I only care about the c++ .exe.

Is there anyone who has uses --nn option and understands the parameters and how they relate to the output prediction?

**Comment by JohnLangford** [comment]
Maybe just disable the C# stuff?  The windows build is somewhat tricky but
it is working routinely in appveyor.


-John

**Comment by apoorvreddy** [comment]
Hi,

I along with @AbdealiJK contributed to the code for fixing the human readable output for the NN.

OutputWeight and HiddenBias are temporary values which are also printed out during the invert_hash procedure. You should ignore them.

Each line in the readable model has the following format:
<feature_name>:<feature_hash>:<feature_weight>

**Weights from input node to the hidden layer:** <feat>, <feat>[1], ... <feat>[k-1]. In your case, your input feature name is 0, and there is only 1 hidden node. So it is just 0:0:-1.06665 in your human readable model.

**Bias to the hidden nodes:** Constant, Constant[1], ... Constant[k-1] (here k is the number of hidden nodes) which are the b1, b2, ... bk. In your case you have 1 hidden node, so the bias to this hidden node is just the Constant value.

**Weights from the hidden layer to the output node:** OutputLayer0[k], OutputLayer1[k], ..., OutputLayerk-1[k] , where k is the number of hidden nodes

**Bias to the Output Node:** Line with OutputLayerConst[k]

Hope this helps ! :)

**Comment by apoorvreddy** [comment]
Regarding @arielf 's experiment. A single layer neural network with a non-linearity cannot possibly learn the identity function i.e f(x) = x . It is possible if the activation function is linear though, which is not the case here.

The reason he gets good accuracy without --nn option is because, it then becomes a simple linear regression problem, where the weights are most probably 1.

**Comment by jp341** [comment]
@apoorvreddy : thanks so much for replying.  

If I understand you correctly, you are suggesting for a 1-node network, predictions f(x) should be in the form:

> f(x) = tanh( x * feat[0] + Constant[0] ) * OutputLayer0[1] + OutputLayerConst[1]

I re-trained my simple VW model above to check if this is true.

- Step 1. Train the Model: 
`vw -d C:\tmp\test_data --nn 1 --passes 100 -k --holdout_off --cache_file C:\tmp\test_model.cache -f C:\tmp\test_model.mod`
- Step 2. Produce Predictions & Readable Model with VW:
`vw -d C:\tmp\test_data -i C:\tmp\test_model.mod --invert_hash C:\tmp\test_model.read -p C:\tmp\test_model.pred`
- Step 3. Produce predictions using f(x) equation above

VW Predictions don't match f(x) calculated predictions.

```
vw  0.000019  f(x) 0.000020        --> error = 0.000001
vw  0.000001  f(x) 0.000020        --> error = 0.000019
vw  0.000019  f(x) 0.000020        --> error = 0.000001
vw  0.000001  f(x) 0.000021        --> error = 0.000020
vw -0.000011  f(x) 0.000020        --> error = 0.000031
vw  0.000005  f(x) 0.000020        --> error = 0.000015
vw  0.000013  f(x) 0.000020        --> error = 0.000007
vw  0.000014  f(x) 0.000020        --> error = 0.000006
```

Feel free to confirm it yourself!  Or have I misunderstood your instructions?

Here is contents of test_data:
```
0.000019 | 0:0.000019
0.000001 | 0:0.000001
0.000017 | 0:0.000017
-0.000011 | 0:-0.000011
0.000005 | 0:0.000005
0.000012 | 0:0.000012
0.000015 | 0:0.000015
0.000018 | 0:0.000018
```

Here is my readable model:
```
:0
0:0:-0.156303
Constant:232120:1.6294e-05
HiddenBias:29015:-0.3161
OutputLayer0[1]:156908:0.233782
OutputLayerConst[1]:156909:1.66623e-05
OutputWeight[1]:156908:0.233782
```
The proposed f(x) to reproduce vw predictions:
`f(x) ( tanh(x * -0.156303 + 1.6294e-5) * 0.233782 + 1.66623e-5 );`

**Comment by apoorvreddy** [comment]
Hi @jp341 , your f(x) seems right !

Are you saying, in

`vw  0.000001  f(x) 0.000020        --> error = 0.000019`

0.000001 is vw prediction ?

**Comment by jp341** [comment]
@apoorvreddy 
Yes, 0.000001 is vw prediction.

Here is the contents of "test_model.pred", the vw output predictions:

```
0.000019
0.000001
0.000019
0.000001
-0.000011
0.000005
0.000013
0.000014
```

**Comment by apoorvreddy** [comment]
I used your test_data
```
0.000019 | 0:0.000019
0.000001 | 0:0.000001
0.000017 | 0:0.000017
-0.000011 | 0:-0.000011
0.000005 | 0:0.000005
0.000012 | 0:0.000012
0.000015 | 0:0.000015
0.000018 | 0:0.000018
```

With the following training command
`vw -d test_data --nn 1 --passes 100 -k -c -f test_model.mod`

And the following test command. You forgot the `-t` flag
`vw -t -d test_data -i test_model.mod --invert_hash test_model.read -p test_model.pred`

This is the human readable file `test_model.read`
```
Version 8.4.0
Id 
Min label:-1.1e-05
Max label:1.9e-05
bits:18
lda:0
0 ngram:
0 skip:
options: --nn 1
Checksum: 3981626792
:0
0:0:-0.717031
Constant:232120:-1.99931e-07
HiddenBias:29015:-0.3161
OutputLayer0[1]:156908:-0.302343
OutputLayerConst[1]:156909:1.454e-05
OutputWeight[1]:156908:-0.302343
```

So `f(x) = tanh(-1.99931e-7 + (-0.717031)*x) * (-0.302343) + 1.454e-5`

This is what I get:
```
Input        VWPred    f(x)
0.000019 0.000018 0.0000187194
0.000001 0.000014 0.0000148172
0.000017 0.000018 0.0000182859
-0.000011 0.000013 0.0000122158
```

The slight difference between VWPred and f(x) is mostly because VW uses an approximate implementation of tanh, called fasttanh in its code.

Hope this clarifies ! :)

**Comment by JohnLangford** [comment]
Thanks @apoorvreddy

**Comment by jp341** [comment]
Thanks @apoorvreddy - you solved my problem! I was omitting the -t flag.

I am now able to reproduce the output of vw from the parameters.

Thanks all.  I will close the ticket.
